### DISCUSSION TITLE
Combining multiple yolo models together with shared backbone and different heads

### DISCUSSION THREAD

**Discussion by yucath** [ultralytics/ultralytics#22141] [General]
Hi, I am trying to combine multiple yolo models (same type - for eg yolov8s-seg, yolov8s-obb, yolov8s). As the models share the common backbone (is that right?), I am trying to separate the backbone and heads of differtn models and tryign to use only one backbone and the different heads (from the full model which was trained separately - and only use head from those). The idea is to reuse the backbone and train wiht more generalised images that can be used across multiple tasks and head is only responsible for some specific classes from the outputs of those backbones. I am exporting backbone adn head separately as .plan files and using triton server to serve the model with a router which tells how to combine those together. 

I am able to do that and get results, but the results do not make sense to me. I think there is something going on. My question is: is this strategy fundamentally wrong? and is there anything that needs to be done to allow for this? 

Open to any suggestions or pointers that can help.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yucath, thank you for starting this thoughtful discussion 🚀 This is an automated response — an Ultralytics engineer will follow up shortly to assist further.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for architecture and usage references where many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples may help frame your approach.

If this is a 🐛 Bug Report related to unexpected behavior when mixing backbones/heads or during export/serving, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question or design discussion, please share:
-Exact model variants and versions used (e.g., yolov8s-seg, yolov8s-obb, yolov8s)  
-Training logs, hyperparameters, and any modifications to model configs  
-Export details (.plan generation steps, toolchain versions) and Triton model repository structure/router configuration  
-Example inputs/outputs that illustrate the mismatch you’re seeing  
-And verify you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Before proceeding, please upgrade to the latest `ultralytics` release and verify all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/). This helps ensure any known issues are already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help us provide targeted guidance on your shared-backbone/different-heads setup with Triton, please include the details above along with any minimal configs or code snippets demonstrating how you’re wiring the backbone output to each head. This will help us quickly validate assumptions and suggest next steps 🔍

**Reply by glenn-jocher** [reply]
Great question. Heads trained separately are strongly co-adapted to their original backbone+neck, so wiring those heads to a different shared trunk usually yields inconsistent outputs. The approach isn’t inherently wrong, but it only works if each head receives exactly the features it was trained on (same P3/P4/P5 order, channels, and strides) and if task-specific branches are honored—segmentation needs prototypes and per-scale mask coefficients, and OBB adds angle logits with rotated-box decoding. The most reliable path is to define a single model with one shared trunk and multiple heads in a custom YAML, train or at least fine-tune the heads jointly on that trunk, then export once to ONNX→TensorRT and serve a single engine with multiple outputs; see Model YAML guidance in [Model YAML Configuration](https://docs.ultralytics.com/guides/model-yaml-config/), head specifics in the [OBB head](https://docs.ultralytics.com/reference/nn/modules/head/) and [YOLO Segmentation head](https://docs.ultralytics.com/reference/nn/modules/head/), how `Detect/Segment/OBB` consume the `ch` feature tuple in the [parse_model reference](https://docs.ultralytics.com/reference/nn/tasks/), and deployment notes in the [Triton guide](https://docs.ultralytics.com/guides/triton-inference-server/).
